apps/src/netsim/NetSimLocalClientNode.js

```diff
@@ -163,76 +175,43 @@ NetSimLocalClientNode.prototype.tick = function (clock) {
 };
 
 /**
- * If a client update fails, should attempt an automatic reconnect.
- * @param {function} [onComplete]
- * @param {boolean} [autoReconnect=true]
+ * Handler for a heartbeat update failure.  Propagates the failure up through
+ * our own "lost connection" callback.
+ * @private
  */
-NetSimLocalClientNode.prototype.update = function (onComplete, autoReconnect) {
-  if (!onComplete) {
-    onComplete = function () {};
+NetSimLocalClientNode.prototype.onFailedHeartbeat_ = function () {
+  logger.error("Heartbeat failed.");
+  if (this.onNodeLostConnection_ !== undefined) {
+    this.onNodeLostConnection_();
   }
-  if (autoReconnect === undefined) {
-    autoReconnect = true;
-  }
-
-  var self = this;
-  NetSimLocalClientNode.superPrototype.update.call(this, function (success) {
-    if (!success && autoReconnect) {
-      self.reconnect_(function (success) {
-        if (!success){
-          self.status_ = 'Offline';
-          self.onChange.notifyObservers();
-        }
-        onComplete(success);
-      });
-    } else {
-      onComplete(success);
-    }
-  });
 };
 
 /**
- * Reconnection sequence for client node, in which it tries to grab a
- * new node ID and propagate it across the simulation.
- * @param {!function} onComplete (success)
- * @private
+ * Give this node an action to take if it detects that it is no longer part
+ * of the shard.
+ * @param {function} onNodeLostConnection
  */
-NetSimLocalClientNode.prototype.reconnect_ = function (onComplete) {
-  var self = this;
-  NetSimLocalClientNode.create(this.shard_, function (node) {
-    if (!node) {
-      // Reconnect failed
-      onComplete(false);
-      return;
-    }
-
-    // Steal the new row's entity ID
-    self.entityID = node.entityID;
-    self.update(function (success) {
-      if (!success) {
-        // Reconnect failed
-        onComplete(false);
-        return;
-      }
+NetSimLocalClientNode.prototype.setLostConnectionCallback = function (
+    onNodeLostConnection) {
+  this.onNodeLostConnection_ = onNodeLostConnection;
+};
 
-      // If we have a wire, we also have to update it to be reconnected.
-      if (self.myWire !== null) {
-        self.myWire.localNodeID = self.entityID;
-        self.myWire.update(function (success) {
-          if (!success) {
-            // Reconnect failed
-            onComplete(false);
-            return;
-          }
+/**
+ * If a client update fails, should attempt an automatic reconnect.
+ * @param {function} [onComplete]
+ */
+NetSimLocalClientNode.prototype.update = function (onComplete) {
+  onComplete = (onComplete !== undefined) ? onComplete : function () {};
```

Comment:
Could simplify as onComplete = onComplete || function () {};

When using this pattern, you need to be careful about falsey values, but since your input is a function, that shouldnt be an issue.